core/reactor.c

```diff
@@ -315,24 +272,18 @@ int next(void) {
 
     LF_PRINT_LOG("Next event (elapsed) time is " PRINTF_TIME ".", next_tag.time - start_time);
     // Wait until physical time >= event.time.
-    // The wait_until function will advance current_tag.time.
     if (wait_until(next_tag.time) != 0) {
         LF_PRINT_DEBUG("***** wait_until was interrupted.");
-        // Sleep was interrupted.
-        // FIXME: It is unclear what would cause this to occur in this unthreaded
-        // runtime since lf_schedule() is not thread safe here and should not
-        // be called asynchronously. Perhaps in some runtime such as for a
-        // PRET machine this will be supported, so here we handle this as
-        // if an asynchronous call to schedule has occurred. In that case,
-        // we should return 1 to let the runtime loop around to see what
-        // is on the event queue.
+        // Sleep was interrupted. This could happen when a physical action
+        // gets scheduled from an interrupt service routine.
```

Comment:
I think that we have to enter a critical section here and reset _lf_timer_interrupted.

Reply:
All we have to know is that an interruption happened, which is already communicated by the return value of `wait_until`... If an interruption happened, we have to peek again to make sure we got the right tag.

Reply:
Oh, you're right about resetting the interrupt flag. This suggests that either we need to make `wait_until` a platform function or we need another function specifically for resetting this flag, which would do nothing for the threaded runtime... We can't just do it in `lf_nanosleep` because the interruption could happen before reaching it (and after leaving the critical section).